Zookeeper setup using offical tarball (also) on debian/ubuntu

lhoss opened the pull request:
final PR to fix: https://github.com/AnsibleShipyard/ansible-zookeeper/issues/26

thx for review @ernestas-poskus 

Update: We could also extract the 2 'configure' tasks from tarball.yml, and re-use the same include file also in the old Debian.yml installation, to further reduce code duplication. Should I do in this PR?

lhoss commented on defaults/main.yml:
```diff
@@ -4,7 +4,9 @@ zookeeper_playbook_version: "0.9.2"
 zookeeper_version: 3.4.6
 zookeeper_url: http://www.us.apache.org/dist/zookeeper/zookeeper-{{zookeeper_version}}/zookeeper-{{zookeeper_version}}.tar.gz
 
+zookeeper_debian_apt_install: false
```
Note: 2 things we could change here:
* set default to 'true'  (assuming we want to keep offering also the original debian apt install)
* rename variable to zookeeper_tarball_install and set 'true' (with the idea we could later also provide a zookeeper yum/rpm setup on redhat)

ernestas-poskus replied:
I would suggest to prefer apt but since 3.4.6 is not available and released - lets leave priority to tarball install

lhoss replied:
I agree.
ps: I suggest to update the documentation a bit (explaining this new setting)
In general I like to have each role variable have a 1-line documentation

lhoss commented on tasks/Debian-tarball.yml:
```diff
@@ -0,0 +1,12 @@
+  tags: deploy
+
+- include: systemd.yml
+  tags: deploy
```
Note: I didn't test the deployment with a later Ubuntu (>15.10 or 16.04 where it's default) which actually runs systemd  (though the systemd template file  looks quite basic, so I'ld expect it to work)

ernestas-poskus replied:
Check 15.04

```bash
vagrant@default-ubuntu-1504:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 15.04
Release:	15.04
Codename:	vivid
vagrant@default-ubuntu-1504:~$ ls -la /usr/lib | grep systemd
drwxr-xr-x  7 root root    4096 Mar 29 21:45 systemd
vagrant@default-ubuntu-1504:~$ ls -la /etc | grep init
drwxr-xr-x  2 root root    4096 Mar 29 21:50 init
drwxr-xr-x  2 root root    4096 Mar 29 21:52 init.d
drwxr-xr-x  5 root root    4096 Mar 29 21:42 initramfs-tools
```

it contains both init and systemd which might lead to unexpected behavior 


15.10 yields the same

```bash
vagrant@vagrant:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 15.10
Release:	15.10
Codename:	wily
vagrant@vagrant:~$ ls -la /etc/ | grep init
drwxr-xr-x  2 root root    4096 Aug  1 16:27 init
drwxr-xr-x  2 root root    4096 Aug  1 16:28 init.d
drwxr-xr-x  5 root root    4096 Aug  1 16:22 initramfs-tools
vagrant@vagrant:~$ ls -la /usr/lib/ | grep systemd
drwxr-xr-x  8 root root    4096 Aug  1 16:24 systemd
```

lhoss replied:
OK, then I suggest to make it configurable for the user, preferably by introducing 1 new variable:
```
# when true systemd will be used (and *not* upstart), when 'false' the other way
zookeeper_debian_systemd_enabled: false (by default)
```

Even better: Set the default setting depending on the ubuntu version (I propose to set it to true only for Ubuntu 16.04, where systemd seems to be the new default init system, and replace upstart)

Alternative we could use 2 separate variables, which would allow to enable/disable both systemd/upstart independently  (could also re-use for redhat, combining those flags with the 'systemd/upstart' exists check ?!)

ernestas-poskus replied:
> Even better: Set the default setting depending on the ubuntu version (I propose to set it to true only for Ubuntu 16.04, where systemd seems to be the new default init system, and replace upstart)

sounds good !

lhoss replied:
> Checked 15.04 ...15.10
> it contains both init and systemd which might lead to unexpected behavior
After some reading I'm quite confident that there shouldn't be a problem to write the configs for both init systems (as only one should be active at the same time).
Ref: https://wiki.ubuntu.com/SystemdForUpstartUsers

Anyway I will now push the proposed+agreed solution, meaning that any user that deviated from the default init system (for ex. when on Ubuntu 15.04 he switched to upstart) has to overwrite the default.

lhoss commented on templates/zookeeper.conf.j2:
```diff
@@ -13,5 +13,5 @@ umask 007
 
 script
-    su -c "{{zookeeper_dir}}/bin/zkServer.sh start-foreground > /var/log/zookeeper/init-zookeeper.log" zookeeper
+    exec sudo -u zookeeper -g zookeeper {{zookeeper_dir}}/bin/zkServer.sh start-foreground > /var/log/zookeeper/init-zookeeper.log 2>&1
```
Note: this change was required to make the upstart script work also in ubuntu (also tested that it still works on centos6+7 !)

lhoss replied:
ps: It is interesting to see  that the 'old' (ubuntu) deb pkg upstart conf file is quite more complex, and completely ignoring zkServer.sh.  Interestingly the logging to file works there.
The script block looks like:
```
    . /etc/zookeeper/conf/environment
    [ -r /etc/default/zookeeper ] && . /etc/default/zookeeper
    if [ -z "$JMXDISABLE" ]; then
        JAVA_OPTS="$JAVA_OPTS -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.local.only=$JMXLOCALONLY"
    fi
    exec start-stop-daemon --start -c $USER --exec $JAVA --name zookeeper \
        -- -cp $CLASSPATH $JAVA_OPTS -Dzookeeper.log.dir=${ZOO_LOG_DIR} \
        -Dzookeeper.root.logger=${ZOO_LOG4J_PROP} $ZOOMAIN $ZOOCFG
end script
```

lhoss commented:
@ernestas-poskus checks passed! happy to get this merged (so I can resolve my 2 other issues, even today) 👍

ernestas-poskus commented on defaults/main.yml:
```diff
@@ -4,6 +4,9 @@ zookeeper_playbook_version: "0.9.2"
 
+# Flag that selects if systemd or upstart will be used for the init service:
+# Note: by default Ubuntu 15.04 and later use systemd (but support switch to upstart)
+zookeeper_debian_systemd_enabled: "{{ ansible_distribution_version|version_compare(15.04, '>=') }}"
```
:clap:

lhoss replied:
Yeah 'version_compare' is a useful filter 👍 
Note that it's only supported since ansible v1.6 (not sure if you mention anywhere a minimal ansible version requirement for the role).
Finally, when running a later test in the provided 'trusty' Vagrant-file/box, that check failed, because ansible was only 1.5.4 (installed by apt)

ernestas-poskus replied:
@lhoss  update this to 1.6 then https://github.com/AnsibleShipyard/ansible-zookeeper/blob/master/meta/main.yml#L7

lhoss replied:
will do!
ps: adding more supported Ubuntu versions (not sure about the existing old(est) ones in that list !)
```
  - name: Ubuntu
    versions:
      - all
...
      - utopic
      - vivid
      - wily
      - xenial
```

ernestas-poskus replied:
IMO lowers Ubuntu version should be utopic 14.10

ernestas-poskus replied:
maybe just change to 1.6

ernestas-poskus commented:
Looks good, I will test this this evening and merge if everything is :ok:. Thank you for contribution !

lhoss commented:
> Looks good, I will test this this evening and merge if everything is 🆗.

Note, with the current state (of the PR), you will notice that there is no zookeeper (log4j) log file 'zookeeper.log' written (as was the case when using the .deb pkg). 
There's however a log written ('init-zookeeper.log') which is the redirected stdout&stderr from (upstart) service script running zkServer.sh  (which is current re-created after each service restart)
Anyway the proper fix (I already implemented & tested) is to enable the 'ROLLINGFILE' log4j appender, which is best done by writing following ENV vars to conf/zookeeper_env.sh.
I've implemented this is using a simple template and a new optional dictionary variable, where you can define your custom ENV settings: 
```
zookeeper_env:
  ZOO_LOG_DIR: "/var/log/zookeeper"
  ZOO_LOG4J_PROP: "INFO,ROLLINGFILE"
```
Notes:
* this also applies for redhat/centos (meaning that till now there was no proper log written here either)
* IMO it would be a reasonable default to enable ROLLINGFILE  (we could do this by setting above env values as a default), except that it would change (improve) also the behavior for the redhat setup
* I would do a separate PR for this (once again waiting for #34 merge)  OR we could add the feature to the current PR ?

ernestas-poskus commented:
Good finding !

> I would do a separate PR for this 

yes this is better option

lhoss commented on meta/main.yml:
```diff
@@ -4,7 +4,7 @@ galaxy_info:
   company: http://michaelhamrah.com
   license: Apache 2
-  min_ansible_version: 1.2
+  min_ansible_version: 1.6
```
@ernestas-poskus done the last change request
